src/main/java/redis/clients/jedis/csc/AbstractCache.java

```diff
@@ -0,0 +1,223 @@
+        if (cacheKeysRelatedtoRedisKey != null) {
+          cacheKeysRelatedtoRedisKey.remove(cacheKey);
+        }
+      }
+      return removed;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public List<Boolean> delete(List<CacheKey> cacheKeys) {
+    lock.lock();
+    try {
+      return cacheKeys.stream().map(this::delete).collect(Collectors.toList());
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public List<CacheKey> deleteByRedisKey(Object key) {
+    lock.lock();
+    try {
+      final ByteBuffer mapKey = makeKeyForRedisKeysToCacheKeys(key);
+
+      Set<CacheKey<?>> commands = redisKeysToCacheKeys.get(mapKey);
+      List<CacheKey> cacheKeys = new ArrayList<>();
+      if (commands != null) {
+        cacheKeys.addAll(commands.stream().filter(this::removeFromStore).collect(Collectors.toList()));
+        stats.invalidationByServer(cacheKeys.size());
+        redisKeysToCacheKeys.remove(mapKey);
+      }
+      stats.invalidationMessages();
+      return cacheKeys;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public List<CacheKey> deleteByRedisKeys(List keys) {
+    if (keys == null) {
+      flush();
+      return null;
+    }
+    lock.lock();
+    try {
+      return ((List<Object>) keys).stream()
+          .map(this::deleteByRedisKey).flatMap(List::stream).collect(Collectors.toList());
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public int flush() {
+    lock.lock();
+    try {
+      int result = this.getSize();
+      clearStore();
+      redisKeysToCacheKeys.clear();
+      getEvictionPolicy().resetAll();
+      getStats().flush();
+      return result;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public Boolean isCacheable(CacheKey cacheKey) {
+    return cacheable.isCacheable(cacheKey.getCommand().getArguments().getCommand(), cacheKey.getRedisKeys());
+  }
+
+  @Override
+  public Boolean hasCacheKey(CacheKey cacheKey) {
+    return containsKeyInStore(cacheKey);
+  }
+
+  @Override
+  public abstract EvictionPolicy getEvictionPolicy();
+
+  @Override
+  public CacheStats getStats() {
+    return stats;
+  }
+
+  @Override
+  public CacheStats getAndResetStats() {
+    CacheStats result = stats;
+    stats = new CacheStats();
+    return result;
+  }
+
+  // End of Cache interface methods
+
+  // abstract methods to be implemented by the concrete classes
+  protected abstract CacheEntry getFromStore(CacheKey cacheKey);
+
+  protected abstract CacheEntry putIntoStore(CacheKey cacheKey, CacheEntry entry);
+
+  protected abstract Boolean removeFromStore(CacheKey cacheKey);
+
+  // protected abstract Collection<CacheKey> remove(Set<CacheKey<?>> commands);
+
+  protected abstract void clearStore();
+
+  protected abstract Boolean containsKeyInStore(CacheKey cacheKey);
+
+  // End of abstract methods to be implemented by the concrete classes
+
+  private ByteBuffer makeKeyForRedisKeysToCacheKeys(Object key) {
```

Comment:
I had to read this ... several times ... and I am still not sure what it does 😄 

Could we simplify the name and put a comment that explains the purpose?

Reply:
convertRedisKeyToBytes how does that sound ?
since you ask, i started to think it needs more attention; now questioning if it is same key when you have a string 'x' and a byte array 'y' ends up with same value.